### ISSUE TITLE
odd thing happening during application error

### ISSUE THREAD

**Issue by cmonacaps** [andywer/threads.js#234]
Hey I put a lot of sauce  - I want to say that maybe the threadsjs code is throwing during an exception.

I've developed an unhealthy obsession with getting my ErrorBoundary right...  But it's displaying something other than the actual exception.

[Lots of views on the problem at this imgur](https://imgur.com/a/BV8lxzt)

Early during application init (but below the Worker and the ErrorBoundary) my code fails to parse the local storage.

I should catch that and use a default value.  But I think there's another problem during the exception handling, that I'm reporting here.

One of the linked images shows how the ErrorBoundary is a child of the Worker manager component.  My common-sense understanding of the word "boundary" led me to hope that putting it like that would prevent whatever is happening in the threadsjs post-error-to-window.

Definitely also possible that [the `ErrorBoundary` or `WorkerManager` component implementations](https://gist.github.com/cmonacaps/2dd39d9b3d7d9b2306b7dd5812484b84) are contributing so including redacted but faithful sketches

Thanks

**Comment by cmonacaps** [comment]
```js
    self.addEventListener("error", event => {
        // Post with some delay, so the master had some time to subscribe to messages
        setTimeout(() => postUncaughtErrorMessage(event.error || event), 250);
    });
    self.addEventListener("unhandledrejection", event => {
        const error = event.reason;
        if (error && typeof error.message === "string") {
            // Post with some delay, so the master had some time to subscribe to messages
            setTimeout(() => postUncaughtErrorMessage(error), 250); 
        }
    });
```

I think it's in this code at the line
`setTimeout(() => postUncaughtErrorMessage(error), 250);`
that it got
> index.js?501f:88 Not reporting uncaught error back to master thread as it occured while reporting an uncaught error already. Latest error: DOMException: Failed to execute 'postMessage' on 'Window': ErrorEvent object could not be cloned.

I think honestly that's the problem, that code needs a fix.

But maybe there's a workaround so I can avoid it getting to this, with some different management of the workers lifetimes?  Lmk :smile:

**Comment by andywer** [comment]
This is kind-of a duplicate of #233. The reason why your ErrorBoundary isn't able to catch that error is that it's an uncaught error in the worker.

I don't see why the line that you quoted would need a fix, but I agree, the error reporting is less than optimal for those edge-casy initialization errors.

Let's discuss in #233 🙂

**Comment by cmonacaps** [comment]
> Let's discuss in #233 🙂

Ok doke thanks for response

Sorry just a tiny bit of incident-specific follow up, probably out of place in the other thread -

> I don't see why the line that you quoted would need a fix, but I agree, the error reporting is less than optimal for those edge-casy initialization errors.

I *was* thinking that the very early application error in one of the nearest children of the worker was causing a race - why 250ms?

But also the exception, `ErrorEvent object could not be cloned`, this is an opaque Error from the React framework, they say in some of the text I got when I called `e.stack` on it (rendered in the ErrorBoundary at the imgur), because they can't know what's happening out of context or something, and I tried calling `Object.getOwnPropertyNames` on it and triggered another exception about how it was missing some basic expected property required by `Object.getOwn**` functions. So I was also guessing this would prevent it from being cloned - probably a clone would use that same method to visit it, since Errors are somewhat opaque usually anyhow.

But now, I wonder if it wasn't this all along.
`Warning (worker-plugin): output.globalObject is set to "window". It must be set to "self" to support HMR in Workers.`

And, I guess if this is how it is, that you can't transport Errors content to and from the workers then perhaps opaque errors are kind of typical and there could be a couple flying, both directions crossing

**Comment by andywer** [comment]
> why 250ms?

Fair question! Really just some quick & dirty timeout that should be long enough to give the master thread time to set up its listeners on 99% of devices out there.

> Failed to execute 'postMessage' on 'Window': ErrorEvent object could not be cloned

I think this is really worker-specific, not React-specific. We could try the following in https://github.com/andywer/threads.js/blob/master/src/worker/index.ts#L178:

```diff
- setTimeout(() => postUncaughtErrorMessage(event.error || event), 250)
+ setTimeout(() => postUncaughtErrorMessage(excludeMethods(event.error || event)), 250)
```

and then define

```js
function excludeMethods(input) {
  if (typeof input === "function") return undefined
  if (typeof input !== "object" || input === null) return null

  if (Array.isArray(input)) {
    return input.map(excludeMethods)
  } else {
    const out = {}
    for (const key in input) {
      out[key] = excludeMethods(input[key])
    }
    return out
  }
}
```

Can you quickly try to monkey-patch it like that? Would be interested to see if that does the job for you.

**Comment by andywer** [comment]
PS: Any chance you can share the error that you saw when using `Object.getOwnPropertyNames()`?

**Comment by cmonacaps** [comment]
Hi just now looking back.  Some things are on a faster cycle 🙂

I'll see if I can reproduce it.  The app is about the same as it was, I moved on since fixing the local storage loader, and that's not hard to break again

Also since then I fixed the warning about the window/self references through some config change.

I'll see how hard it is to reproduce and try the patch, then update here

> the error that you saw when using Object.getOwnPropertyNames()

will update

**Comment by cmonacaps** [comment]
I can reproduce the problem pretty easily alright, but then nothing's easy after that.

Only real problem for my application is I need to load local storage safely. Because the longer I look at it I'm pretty sure the error hit by the workers is an effect of the application faulting so soon, perhaps during their initialization.

I'll step through this, but I couldn't figure it out as yet:

I tried a few approaches editing or injecting into the export from worker/index. Did not find a way.  Just going off what's in my node_modules, looks like I have the dist-esm flavor 1.4.0.

I forked threads.js and that built and tested very easily, then after setting an additional resolve.modules path in webpack that was getting used, but I found that my worker did not get any exports from it.  I'm not sure why but I took a quick guess the threads-plugin had to also know where to get it and that was more of a challenge to get set up.

After an hour or two I abandoned that effort and went for just hacking on the worker/index.js under node_modules.  That was viable.  I'll just step through this now because I haven't been able to make sense of it as yet.

basic repro
```
Not reporting uncaught error back to master thread as it occured while reporting an uncaught error already. Latest error: DOMException: Failed to execute 'postMessage' on 'Window': ErrorEvent object could not be cloned.
    at Object.postMessageToMaster (webpack-internal:///./node_modules/threads/dist-esm/worker/implementation.browser.js:8:10)
    at postUncaughtErrorMessage (webpack-internal:///./node_modules/threads/dist-esm/worker/index.js:99:65)
    at eval (webpack-internal:///./node_modules/threads/dist-esm/worker/index.js:198:15)
```

added `excludeMethods` and inserted as indicated
```
Uncaught RangeError: Maximum call stack size exceeded
    at Function.isArray (<anonymous>)
    at excludeMethods (index.js:171)
    at excludeMethods (index.js:176)
    at excludeMethods (index.js:176)
    at excludeMethods (index.js:176)
    at excludeMethods (index.js:176)
    at excludeMethods (index.js:176)
    at excludeMethods (index.js:176)
    at excludeMethods (index.js:176)
    at excludeMethods (index.js:176)
```
where 171 and 176 refer to
```js
/* 171 */  if (Array.isArray(input)) {
             return input.map(excludeMethods)
           } else {
             const out = {}
             for (const key in input) {
/* 176 */      out[key] = excludeMethods(input[key])
```

I tried two things at this point not worth reproducing here, I added a parent reference parameter to excludeMethods so it would additionally pass `input` when invoking itself on an element of `input`, and put an early return with the others at top of the function `if (input === parent) return`.  When that didn't do anything I tried changing the `parent` parameter to a `parents` stack and the early return was `if (-1 === parents.indexOf(input)) return`, also didn't really lead to much - still faulting repetitively  and still unclear why.

so I stopped trying to check if it's a recursive cycle recursively and set up the parent/child constraint before recurring:
```js
  if (Array.isArray(input)) {
    return input.map(el => el === input ? null : excludeMethods(el))
  } else {
    const out = {}
    for (const key in input) {
      out[key] = input[key] === input ? null : excludeMethods(input[key])
    }
    return out
  }
```
not much difference:
```
Uncaught RangeError: Maximum call stack size exceeded
    at excludeMethods (index.js:176)
    at excludeMethods (index.js:176)
    at excludeMethods (index.js:176)
    at excludeMethods (index.js:176)
```
and this is at
```
    for (const key in input) {
      out[key] = input[key] === input ? null : excludeMethods(input[key])
                      ^                        ^
                      first frame              subsequent frames
```
    
then I added a bit of console.log to the method:
```js
function excludeMethods(input, depth=0) {
  if (typeof input === "function") return undefined
  if (typeof input !== "object" || input === null) return null

  if (Array.isArray(input)) {
    return input.map(el => el === input ? null : excludeMethods(el))
  } else {
    const out = {}
    for (const key in input) {
      console.log(depth, typeof input, typeof key, typeof input[key]);
      out[key] = input[key] === input ? null : excludeMethods(input[key], depth+1)
    }
    return out
  }
}
```

Here the output is a bit more interesting as the trace of faults did not change much at all.
<img width="410" alt="Screen Shot 2020-04-19 at 9 40 49 PM" src="https://user-images.githubusercontent.com/50930812/79706546-9e860300-8287-11ea-852c-a9da6294b330.png">
I'm using the screenshot to convey this since it includes the Chrome console "group similar" feature - the count of grouped lines does not get picked up to the clipboard with the text but it's easy to see the repeated pattern in the grouping counts.  It got to a depth of about 5900 before it got halted and this pattern is consistent insofar as I can still scroll to it, roughly "19 2 pass 2 8 4 pass 4 7 3 4 ..."

try seeing the strings,
```js
    for (const key in input) {
      console.log(depth, typeof key === "string" ? `'${key}'` : typeof key,
                         typeof input[key] === "string"
                                  ? input[key]
                                  : typeof input[key]);
      out[key] = input[key] === input ? null : excludeMethods(input[key], depth+1)
    }
```

I dunno, looks like we're trying to clone the window:
<img width="805" alt="Screen Shot 2020-04-19 at 10 00 48 PM" src="https://user-images.githubusercontent.com/50930812/79707079-4d770e80-8289-11ea-9cd7-303dec57c192.png">

😄

**Comment by cmonacaps** [comment]
Sorry, no idea why it would be doing that.

```js
function excludeMethods(input, depth=0) {
  console.log(typeof self, typeof input);
  if (input === self) {
    console.log("input is self, stop");
  }
  if (depth > 100) { return; }
  if (typeof input === "function") return undefined
  // ...
```
<img width="831" alt="Screen Shot 2020-04-19 at 10 12 43 PM" src="https://user-images.githubusercontent.com/50930812/79707565-f07c5800-828a-11ea-964d-ecd31fe39d79.png">

So near the beginning, I hit that "self, stop" rule and I guess that was enough to stop digging down one branch but not to stop it entirely.

I'm not sure, I think I'll stop here for a bit for now.

**Comment by andywer** [comment]
Wow!

Try using `threads@1.4.0-error-reporting-fix` to get a better look at the initialization error (see #233) 😉

**Comment by cmonacaps** [comment]
Thanks! will try that! 😄

**Comment by cmonacaps** [comment]
hi just touching base to verify the issue's resolved, thanks again! 👍

**Comment by andywer** [comment]
Great! 🎉Just released the fix as version `1.4.1`.
